### DISCUSSION TITLE
I want to simulate pouring objects into a cup

### DISCUSSION THREAD

**Discussion by miyukin73** [google-deepmind/mujoco#357] [Asking for Help] [answered]
Hello,
I'd like to simulate pouring objects into a cup.

So, I created the STL file of the cup in CAD.
![image](https://user-images.githubusercontent.com/63595993/175295008-5452e008-0731-4aa5-a795-b78efae101bd.png)

However, I found the following text at the URL site below.

" if you want to model a non-convex shape, you have to decompose it into a union of convex geoms and attach all of them to the same body. "
https://mujoco.readthedocs.io/en/latest/overview.html

In fact, using the STL file of the created cup (black object), the object (green sphere) did not fit into the cup as shown in the attached photo.

![image](https://user-images.githubusercontent.com/63595993/175294879-40f1fec7-f51e-425b-8258-13614fa96289.png)

I saw the answer to question "hollow cynlinder geom #19", 
but as a beginner, I don't know how to convert from concave to convex and would like to know.

Thank you!
[pouring objects.zip](https://github.com/deepmind/mujoco/files/8967028/pouring.objects.zip)

**Comment by kevinzakka** [comment, marked as answer]
Hi @miyukin73, you can use [V-HACD](https://github.com/kmammou/v-hacd) to decompose the mesh into approximate convex submeshes. I've made a tool to simplify this process [here](https://github.com/kevinzakka/obj2mjcf). Run the [bash script](https://github.com/kevinzakka/obj2mjcf/blob/main/install_vhacd.sh) to install V-HACD and then you can run the V-HACD executable on your mesh. Note you'll want to convert the STL to an OBJ file to use it with V-HACD, you can do that with either Blender or [trimesh](https://github.com/mikedh/trimesh).

**Reply by miyukin73** [reply]
Hello, @kevinzakka 

Thank you so much for your prompt and kind response.
However, I'd like to ask you something that I still do not understand.

I have done the following:
・git clone https://github.com/kevinzakka/obj2mjcf.git
・pip install --upgrade obj2mjcf
・bash install_vhacd.sh
・git clone https://github.com/mikedh/trimesh.git
・pip install trimesh
・pip install trimesh[easy]

I then tried to run the "Quick Start" program, but got the following error:
AttributeError: module 'trimesh' has no attribute 'util'

What could possibly go wrong?
Also,  I don't understand how to "run the V-HACD executable on your mesh".

I'm very sorry for the basic information, but I would like to know.
Thank you very much.

**Reply by kevinzakka** [reply]
Hi @miyukin73,

You can simply do the following to convert an STL file to an OBJ file:

```python
import trimesh

mesh = trimesh.load("cup.stl")
mesh.export("cup.obj", file_type="obj")
```

Given the obj file, you can then do:

```bash
obj2mjcf --obj-dir . --obj-filter cup --save-mjcf --compile-model --vhacd-args.enable
```

Let me know if you have any issues!

**Reply by miyukin73** [reply]
Hello @kevinzakka 

Thank you very much for your reply.

I'm sorry but  I got the following error:
AttributeError: partially initialized module 'trimesh' has no attribute 'load' (most likely due to a circular import)

How should I respond?
Sorry to keep repeating myself, but thank you in advance.

**Reply by kevinzakka** [reply]
Are you inside the trimesh directory or something? If not, can you pip uninstall trimesh and do a clean install?

**Reply by miyukin73** [reply]
Thank you very much for your  instruction, @kevinzakka !

 I ran all the code you gave me and created a new folder in trimesh called "cup" as attached!

However, I am not sure what to do in the next step.

I originally imported cup.stl in the attached jaco.xml file, but how do I use the cup file I just created?

Sorry for the inconvenience, but any hints would be appreciated.

Thank you.

[cup.zip](https://github.com/deepmind/mujoco/files/8969910/cup.zip)
[jaco2.zip](https://github.com/deepmind/mujoco/files/8969909/jaco2.zip)

**Reply by kevinzakka** [reply]
You can either include the cup file in the jaco file or copy paste the relevant meshes / textures / geoms into the jaco file. Whatever works for your current workflow.

**Reply by miyukin73** [reply]
Thank you very much @kevinzakka 

I  transcribed the contents of cup.xml into jaco.xml.

However, when I run it, I get the following error
![image](https://user-images.githubusercontent.com/63595993/175375014-32c3249d-9edc-4d3a-ace0-ade5a94196e8.png)

![image](https://user-images.githubusercontent.com/63595993/175375107-7b6a514b-4072-4d88-9b04-08af37e81fa7.png)

Is there any solution to this problem?

I would appreciate it if you could reply when you have time.
Thank you.

**Reply by kevinzakka** [reply]
Not 100% but looks like you're using mujoco_py? Unclear what MuJoCo version it uses, but OBJ support was only added in [2.1.2](https://github.com/deepmind/mujoco/releases/tag/2.1.2).

**Reply by miyukin73** [reply]
I used mujoco-2.1.5 and was able to run the simulation successfully!

Thank you very much for your very kind support!!

**Reply by miyukin73** [reply]
Hello, @kevinzakka 

I am very sorry for asking the same question.
I have been away for a little while, but I tried to run the following program that you taught me before to convert from stl to obj files again.

```
import trimesh

mesh = trimesh.load("cup.stl")
mesh.export("cup.obj", file_type="obj")
```

However, the following error occurred.

Data not utf-8! Trying ISO-8859-1 (confidence 0.73)

I feel that it is not a problem with the stl file because the same error occurred with the file that worked before.
What is the problem?
I would appreciate it if you could tell me.
Thank you in advance.

**Comment by miyukin73** [comment]
Hello @kevinzakka 

I am sorry for the frequent interruptions...

I was successful if I just run the XML file using[ mujoco-2.1.5.](https://github.com/deepmind/mujoco/releases/tag/2.1.5)
However, when I run the same XML file using [mujoco_py](https://github.com/studywolf/mujoco-py/), I get the error below.

![image](https://user-images.githubusercontent.com/63595993/175517568-7959b6d6-5612-4864-8f7e-a6ec787281a6.png)


As far as I confirm, mujoco_py was updated in November 2021 and MuJoCo was updated in May 2022, 
so maybe mujoco_py has not kept up with updates...?

Anyway, I would like to ask if you know of any solution to this problem.

Thank you.

**Comment by yuvaltassa** [comment]
Hi,

I made you an example model. Copy this file into `models/mug/` and load it

<details>
  <summary>mug with balls</summary>

```XML
<mujoco model="mug with balls">
  <include file="mug.xml"/>
  <size nconmax="1000" njmax="3000"/>
  <worldbody>
    <light pos="0 0 1" diffuse=".3 .3 .3"/>
    <composite type="particle" count="5 5 5" spacing="0.011" offset=".005 .003 .15">
      <geom size=".005" rgba=".8 .2 .1 1" solref=".005 1"/>
    </composite>
  </worldbody>
</mujoco>
```

</details>

This is what it look like:

https://user-images.githubusercontent.com/6387526/175768374-374e7e5d-4b93-466e-868f-4675a0f5c992.mp4

**Reply by miyukin73** [reply]
Hello @yuvaltassa 

Thank you so much for taking time out of your busy schedule to create the example model!
I wanted to create a simulation environment just  like the one in the movie...!!!

However, I also got the error below when using mujoco-2.1.5.
![image](https://user-images.githubusercontent.com/63595993/175823843-e5585f96-15ff-462d-95f6-8e4cc68d1487.png)

I am really sorry to trouble you with something so basic, but I am new to mujoco and don't know how to deal with it, 
so I would like to know.
Thank you very much.

**Reply by yuvaltassa** [reply]
It says FILE_NOT_FOUND, your file is not where you think it is...

**Reply by miyukin73** [reply]
Hello @yuvaltassa 

Sorry... I didn't activate the conda...
Thank you very much for your advice.

However, when I tried to run the XML file in the environment again, I got the following error.
![image](https://user-images.githubusercontent.com/63595993/175861788-eb9ecae4-e1a3-4419-addd-5b6d4b82721f.png)

Do you know what the cause is...?
Thank you.

**Reply by yuvaltassa** [reply]
Can you try reducing `shadowsize` (line 4 in mug.xml)? Try 1024 and let me know what happens.

**Reply by miyukin73** [reply]
Thank you very much for your advice, @yuvaltassa !

I reduced shadowsize, and the XML file is successfully executed!!

However, I have two additional questions.

The first question is regarding the video you sent me, in the video it looked like you were clicking and knocking over the cup, 
how can that be done...?

The second question is regarding mujoco_py, 
I still get the error below when I use mujoco_py. is it a problem with the version of mujoco_py that it cannot read obj files?

![image](https://user-images.githubusercontent.com/63595993/175943583-f7e8bebb-6882-4f05-9caa-f5e03f7dd24f.png)


I am sorry to keep repeating this, but I would like to know.
Thank you in advance.

**Comment by yuvaltassa** [comment]
> in the video it looked like you were clicking and knocking over the cup, how can that be done...?

I'm using `simulate`, the [built-in native viewer](https://mujoco.readthedocs.io/en/latest/programming.html#sasimulate). Hit 'F1' for basic help and take it from there. I think specifically for this you want double-click to select the cup, then control-drag to apply a torque.

> I still get the error below when I use mujoco_py. is it a problem with the version of mujoco_py that it cannot read obj files?

Sorry, we don't support mujoco-py. It is an unmaintained package. Please switch to the native bindings and/or `dm_control`.

**Reply by miyukin73** [reply]
Thanks for letting me know, @yuvaltassa .
could over the cup using simulate!

I understand.
I would like to ask you if you know anything about `dm_control.`
I want to use python code to display a Jaco2 arm robot and the mug in the simulation environment, 
and then Jaco2 grabs the mug and tilts it.
![image](https://user-images.githubusercontent.com/63595993/175965352-d02777c2-f079-40ce-bdb5-5fa394207ad7.png)

Is this possible with dm_control?

**Reply by miyukin73** [reply]
Hi @yuvaltassa 
I’m sorry for repeating myself.

Or, I would like to do something like rotating a mug at a constant angular acceleration, like I did simulate.
Are there any sample programs?
